### PR TITLE
ASoC: SOF: ipc4-topology: Fix logic for input format selection for DA…

### PR DESCRIPTION
**Pull request by ranj063**
…I copier capture

Check if all available input audio formats for a DAI copier have the same sample format and use that for reference when selecting the input format. This is needed for the case when a DAI copier for the DMIC supports only 32-bit capture but has multiple input audio formats in topology in order to support different channel counts and rates.

### CODE REVIEW

**Comment by plbossart** [comment]
@ranj063 can you remind me why this is needed and what this fixes? We just merged DMIC related stuff, no following if this is related or a new problem?

**Comment by ranj063** [comment]
> @ranj063 can you remind me why this is needed and what this fixes? We just merged DMIC related stuff, no following if this is related or a new problem?

@plbossart the DMIC PR that was merged was to deal with NHLT blobs. Typically in our topologies and the BIOS NHLT, we only have 32-bit format blobs and in order to support 16-bit capture, the PR proposed using the 32-bit blob so that we can use 32-bit format for the DMIC DAI copier and convert to 16-bit at the host copier.

But apart from this , the other problem is that our topology has multiple input formats for the DMIC DAI copier in order to support both 2ch and 4ch capture even though the input sample format in both cases in 32-bit. So, in this case, we need to restrict the reference sample format to 32-bit when trying to find a match for the input.

**Comment by plbossart** [comment]
> > @ranj063 can you remind me why this is needed and what this fixes? We just merged DMIC related stuff, no following if this is related or a new problem?
> 
> @plbossart the DMIC PR that was merged was to deal with NHLT blobs. Typically in our topologies and the BIOS NHLT, we only have 32-bit format blobs and in order to support 16-bit capture, the PR proposed using the 32-bit blob so that we can use 32-bit format for the DMIC DAI copier and convert to 16-bit at the host copier.
> 
> But apart from this , the other problem is that our topology has multiple input formats for the DMIC DAI copier in order to support both 2ch and 4ch capture even though the input sample format in both cases in 32-bit. So, in this case, we need to restrict the reference sample format to 32-bit when trying to find a match for the input.

You left me in the dust, I have no idea what this is about, sorry.

**Comment by singalsu** [comment]
@plbossart We want to keep capture DAI 32 bits even if capture format to user space is 16 bits. We can't provide to user space 16 bit quality with 16 bit pre-processing pipelines since they all add quantization noise (iir/dcblock, volume, in future more processing). With 32 bits pipelines the noise floor due to processing is much less than 16 bits. Also if need to add gain to capture like in DMIC case or similar lower sensitivity (and high-SPL capable) analog microphone we need the extra precision bits to "shift" to upper parts of the word.

Earlier version of this patch plus now merged #4840 has worked for me to fix the 16 bit capture fail.

**Comment by plbossart** [comment]
> @plbossart We want to keep capture DAI 32 bits even if capture format to user space is 16 bits. We can't provide to user space 16 bit quality with 16 bit pre-processing pipelines since they all add quantization noise (iir/dcblock, volume, in future more processing). With 32 bits pipelines the noise floor due to processing is much less than 16 bits. Also if need to add gain to capture like in DMIC case or similar lower sensitivity (and high-SPL capable) analog microphone we need the extra precision bits to "shift" to upper parts of the word.
> 
> Earlier version of this patch plus now merged #4840 has worked for me to fix the 16 bit capture fail.

Okay, this makes sense @singalsu, but the commit message doesn't even begin to explain how this is related to 16 bits and that the choice of 32 bits is intentional to improve headroom/avoid quantization.

Honestly this looked like ChatGPT-speak with the prompt: "generate a commit message that will confuse reviewers" :-)

**Review by singalsu — APPROVED** [review]
Thanks, this fixes the problem. I just checked with fresh sof-dev and this patch. Now I can do with a test topology:

```
arecord -Dhw:0,6 -f S32_LE -c 4 -r 48000 ...
arecord -Dhw:0,6 -f S16_LE -c 4 -r 48000 ...

```
and both work correctly. Previously the latter failed.

**Comment by singalsu** [comment]
> Okay, this makes sense @singalsu, but the commit message doesn't even begin to explain how this is related to 16 bits and that the choice of 32 bits is intentional to improve headroom/avoid quantization.
> 
> Honestly this looked like ChatGPT-speak with the prompt: "generate a commit message that will confuse reviewers" :-)

Yep, the problem statement is missing or well hidden from commit text.  I pressed already accept but good to improve it.

File: sound/soc/sof/ipc4-topology.c [hunk 1]
```diff
@@ -1589,8 +1590,48 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
...
+
+		m = hw_param_mask(&ref_params, SNDRV_PCM_HW_PARAM_FORMAT);
+		snd_mask_none(m);
+		snd_mask_set_format(m, snd_fmt);
+	}
+
 	/* set input and output audio formats */
-	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &copier_data->base_config, ref_params,
+	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &copier_data->base_config, &ref_params,
 					    available_fmt);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c:1668]: I don't get how this works...
When we have single format, this all looks like a NOP: the sof_ipc4_init_input_audio_fmt() ignores the ref_params in that case, so the whole `if (single_input_format && dir == SNDRV_PCM_STREAM_CAPTURE)` is just done for nothing.
Line 1199

I'm sure I'm missing something...

We try to narrow the  format support if only one format is supported (which should by itself should narrow it?), but what is expect if we have two formats supported, like S24_LE and S32_LE? We are no masking out the S16_LE?

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -1589,8 +1590,48 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		return -EINVAL;
 	}
 
+	/* check if all input formats are the same */
+	single_input_format = sof_ipc4_copier_is_single_format(sdev,
+								available_fmt->input_pin_fmts,
+								available_fmt->num_input_formats);
+
+	/*
+	 * in the case of capture, if all the input formats are the same, update the reference
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c:1632]: So, if all the input formats are the same then why we have multiple of them? Same format, but different rate?

**Review by ujfalusi — COMMENTED** [review]
I think I'm lost with this with the description of
if all the input formats are the same (why we have multiple if they are the same?) then select from the formats (from the identical ones) the one which is the only one?

**Comment by ranj063** [comment]
> I think I'm lost with this with the description of if all the input formats are the same (why we have multiple if they are the same?) then select from the formats (from the identical ones) the one which is the only one?

@ujfalusi what I mean is multiple input audio format objects in topology but with different channels/rates but same sample format

**Comment by ranj063** [comment]
> > @plbossart We want to keep capture DAI 32 bits even if capture format to user space is 16 bits. We can't provide to user space 16 bit quality with 16 bit pre-processing pipelines since they all add quantization noise (iir/dcblock, volume, in future more processing). With 32 bits pipelines the noise floor due to processing is much less than 16 bits. Also if need to add gain to capture like in DMIC case or similar lower sensitivity (and high-SPL capable) analog microphone we need the extra precision bits to "shift" to upper parts of the word.
> > Earlier version of this patch plus now merged #4840 has worked for me to fix the 16 bit capture fail.
> 
> Okay, this makes sense @singalsu, but the commit message doesn't even begin to explain how this is related to 16 bits and that the choice of 32 bits is intentional to improve headroom/avoid quantization.
> 
> Honestly this looked like ChatGPT-speak with the prompt: "generate a commit message that will confuse reviewers" :-)

@plbossart ive updated the commit message now. Let me know if it still looks like chatgpt-generated.

**Comment by plbossart** [comment]
Looks better @ranj063, but it can be improved further. The problem is that you use 'format' in two separate contexts

"
In topology the DAI copier may
have multiple input audio format objects that have different channel
counts/rates with the same 32-bits for sample format across all of them.
"

the input audio format object contains a sample format?

Can we try to find another term so that 'format' is used in a unique way.

**Comment by fredoh9** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
> Looks better @ranj063, but it can be improved further. The problem is that you use 'format' in two separate contexts
> 
> " In topology the DAI copier may have multiple input audio format objects that have different channel counts/rates with the same 32-bits for sample format across all of them. "
> 
> the input audio format object contains a sample format?
> 
> Can we try to find another term so that 'format' is used in a unique way.

I've replaced sample format with bit depth.

**Review by plbossart — APPROVED** [review]
Thanks @ranj063, this is clear-enough for me.

**Review by ujfalusi — APPROVED** [review]
@ranj063, this is heavy ;)

We should mention that this _must_ accompany https://github.com/thesofproject/linux/pull/4840 ?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, I think this and #4840 should be combined as a fixup for `ASoC: SOF: ipc4-topology: Look for the 32-bit DMIC blob first`

Basically drop the `ASoC: SOF: ipc4-topology: Look for the 32-bit DMIC blob first` and do the `ref_params` refinement in case of single input <strike>format</strike> bit_depth earlier, that will pass the only available params_width to snd_sof_get_nhlt_endpoint_data()

If we don't have correct blob for the only supported depth then we anyway going to fail, no?
I mean: if we only have blob for 16bit but the copier only have support for 32bit then what we will should do?

Atm, we would use t he 16bit blob but a 32bit copier config, is that going to work?

The problem is that we use two 'conditions' for the two things which depend on each other and there could be a disconnect in the configuration.

**Comment by ranj063** [comment]
> @ranj063, I think this and #4840 should be combined as a fixup for `ASoC: SOF: ipc4-topology: Look for the 32-bit DMIC blob first`
> 
> Basically drop the `ASoC: SOF: ipc4-topology: Look for the 32-bit DMIC blob first` and do the `ref_params` refinement in case of single input format bit_depth earlier, that will pass the only available params_width to snd_sof_get_nhlt_endpoint_data()
> 
> If we don't have correct blob for the only supported depth then we anyway going to fail, no? I mean: if we only have blob for 16bit but the copier only have support for 32bit then what we will should do?
> 
> Atm, we would use t he 16bit blob but a 32bit copier config, is that going to work?
> 
> The problem is that we use two 'conditions' for the two things which depend on each other and there could be a disconnect in the configuration.

@ujfalusi these 2 patches are both requirements for 16-bit capture to work but I'm not sure squashing them up together makes sense. They address 2 separate problem areas and I think it makes sense to keep them separate as they fix 2 different kinds of errors

**Comment by ujfalusi** [comment]
@ranj063, this is what I was thinking of: #4899

**Comment by ranj063** [comment]
replaced by #4899

**Pull request closed without merging**